### PR TITLE
Destroy all adjustments on cart empty

### LINKED ISSUE #9112: [Paypal & Stripe] Failed checkout attempts add a persistent transaction fee
## Description

If a Paypal transaction fails on checkout and a transaction fee is associated to that payment, the respective amount will be added to that order, as many times as checkout attempts (with Paypal).

This leads to an inconsistent state of the order / fail cascade, in which:

- checking out with cash is possible, but the order total will contain those ghost transaction fees :warning: :
- checkout with Stripe will explode :boom: : display (the warning below) and resets the cart which is a good thing. It also triggers this bugsnag error:

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/61ebce343d0f08000872149b?event_id=625e8a3d0093606c2ee10000&i=sk&m=nw

## Expected Behavior

Failed checkout attempts using Paypal should not add a persistent transaction fee.

## Actual Behaviour

Failed checkout attempts using Paypal should not add a persistent transaction fee.

## Steps to Reproduce

As Admin:
1. Set up a transaction fee on a Paypal payment method
2. Edit the Paypal settings: for a valid `Signature` add an empty space after it :eyes: 

As Customer:
3. Checkout on that shop, selecting that payment method
4. Notice the error on checkout
5. Check your cart, and notice the transaction fee was added
6. Empty your cart, and start over from step 3.
7. Notice the previously added transaction fee is still on the cart: transaction fees add up...
8.1 Complete this funky order using Stripe -> :boom: -> Cart should be reset.
8.2 8.1 Complete this funky order using Cash -> Order is completed with the ghost transaction fees included in the order total :warning:  

## Animated Gif/Screenshot

Peek 2022-04-19 11-06

## Workaround

There are some ways to deal with the bug, but it can go unnoticed.
I'd say this is an S2 but feel free to change if you disagree.

## Severity

bug-s2: a non-critical feature is broken, no workaround

## Your Environment

* Version used: v4.1.26
* Browser name and version: Firefox 99
* Operating System and version (desktop or mobile): 20.04

### PR DESCRIPTION
**Pull request by georgethoppil**
#### What? Why?

Closes #9112 

#### What should we test?
Scenario 1
- Set up local env and shop to accept credit card payment. Add an extra space in the signature 
- Pay an order with credit card and get an error
- Empty cart
- Pay a new order with cash. Make sure you don't see a ghost transaction fee

Scenario 2
- Set up local env and shop to accept credit card payment. Add an extra space in the signature 
- Pay an order with credit card and get an error
- Empty cart
- Pay a new order with credit card. Make sure you don't see multiple transaction fees 

#### Release notes

Changelog Category: Technical changes

The title of the pull request will be included in the release notes.

### CODE REVIEW

**Review by mkllnk — APPROVED** [review]
Good find.

**Comment by mkllnk** [comment]
@georgethoppil, please assign yourself to all your pull requests.

**Comment by drummer83** [comment]
Hi @georgethoppil and thanks for this PR as well!


I was NOT able to reproduce the bug with **credit cards** in master, even using the settings provided here: https://github.com/openfoodfoundation/openfoodnetwork/issues/9112#issuecomment-1106439837
Using a **failing credit card payment** (card 4000 0000 0000 0002) and a **flat percent** payment fee calculator:
- no ghost transaction fees are applied and the second payment is correct (fee applied only once) for cash and credit card payment.

However using a **failing PayPal payment**, there are still some issues!
Scenario 1:
- After the PayPal payment failed, there is already a duplicated transaction fee applied to the cart summary (on the very right in the screen capture)
- Emptying the cart and then placing a new order with cash payment is failing with a snail.

Scenario 2:
- After the PayPal payment failed, there is already a duplicated transaction fee applied to the cart summary (on the very right in the screen capture)
- Emptying the cart and then placing a new order with credit card payment is failing with a snail.


### Summary
I don't see any change in behavior at all.
@filipefurtad0 should help me reproduce the bug with credit cards.
The PayPal issues should be fixed before merging.

Moving back to In Dev.

**Comment by georgethoppil** [comment]
@drummer83 Thanks for testing it out 
It's weird you are not able to reproduce the issue with the ghost transaction. Hopefully @filipefurtad0 will be able to check in on that. 

For both those scenarios that you tested, did it also blow up when you were on the master branch? I would assume so since the code changes in this pr is very minimal and don't expect those changes to cause that crash. If thats true and it is independent of this pr, it should be raised as another ticket imo. None the less, were there any logs for those crashes in your network manager? Maybe its a quick fix

**Comment by georgethoppil** [comment]
So when we empty the cart, it also destroys the shipments associated with the order. 

The checkout process is done via state transitions. So during our first checkout, we go through all the state transitions and we are at the last state of payment. When we empty the cart, we delete the shipments for the order and then we add the new items to cart. We then go back to checkout but it is already at the last transition of state which is the payment. We dont create the shipments records here which is the problem. We don't really reset the checkout transition after we do the empty cart. I added this now and it seems to work for me. 
@mkllnk probably need another review on this if that makes sense

**Comment by mkllnk** [comment]
That sounds totally reasonable. Just the specs need some updates because they work with in-memory models and the checkout works only with persisted models.

**Comment by drummer83** [comment]
Hi @georgethoppil, thanks for looking at it again.

It's improving but we still have issues with failing PayPal payments.
Scenario 1:
- After the PayPal payment failed, there is already a duplicated transaction fee applied to the cart summary (on the very right in the screen capture) **--> this is still the case and needs fixing**
- The transaction fee is still in the cart (see screen capture) **--> this needs fixing**
- Emptying the cart and then placing a new order with cash payment is failing with a snail. --> **this now working** (see screen capture) :partying_face: 
Peek 2022-05-29 13-15

Scenario 2:
- See Scenario 1: No snail anymore, but still added transaction fee (which is not actually added to the amount due, see below).

The good thing is transaction fees are not piling up anymore. :muscle: 
BUT: There is still the first transaction fee ending up in the cart and showing in the order summary, even if cash (no fee) is selected. 
The good thing again is that it's not really added to the amount due (the order confirmation is showing the correct amount without transaction fee).
Here is a screen capture of this scenario:
Peek 2022-05-29 13-26

**My Proposal:**
1. I would like to ask @filipefurtad0 to test the failing Stripe payment on this issue (as I could not reproduce).
2. Both (Stripe and PayPal) need to be tested with split checkout activated.
3. This PR is an improvement (no snails, no piling up of fees and the correct amount is charged) so if 1. and 2. are ok, we could merge and
4. Create a new issue to solve the remaining problem of previous transaction fees showing up in the cart and in the order summary.

Thanks!

**Comment by filipefurtad0** [comment]
Hey @georgethoppil ,

We're currently [unable to deploy this PR](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/servers/uk-staging/deploys/1231) for further testing due to some `@storybook/addons@6.5.6\" has unmet peer dependency`

This is probably due to some recent changes on master. Could you please rebase the PR?

Thank you :pray:

File: app/models/spree/order.rb [hunk 1]
```diff
@@ -433,6 +433,7 @@ def empty!
       all_adjustments.destroy_all
       payments.clear
       shipments.destroy_all
+      restart_checkout_flow if state == "payment"
```
Comment (filipefurtad0) [inline app/models/spree/order.rb:436]: Have not tested yet - will stage deploy right after this comment - but this makes sense to me :+1: 

Also, we'll probably need the same approach to address #9131 , but with state confirmation in mind. Anyway, something for a separate PR.

**Review by filipefurtad0 — APPROVED** [review]

**Comment by filipefurtad0** [comment]
Hey @georgethoppil and @drummer83,

On **legacy checkout**
After going through scenarios 1 and 2, on the testing notes it was possible to confirm that:
- checkout is possible with other methods other than Paypal :+1:
- "ghost" transaction fees are cleared up after the user visits the `/cart` page and clicks **empty cart** option.

On **split checkout**
- After going through scenarios 1 and 2 we end up with an order in `state = confirmation` and no shipments, which then prevents the rendering of the `/confirmation` page. The PR does not change this behavior, but the original issue was rather about the legacy checkout anyway.

After we test split checkout, if we then toggle the legacy checkout again then the first payment with **stripe will show a snail** - although the order is completed. I think this is rather related to the state of the order, and not really relevant in the context of this PR which only changes the behavior of the "empty cart" option.

This is a great improvement as it prevents adjustments from one order leaking to other orders. So, I'd merge and close #9112.

As to your findings @drummer83 : 

> After the PayPal payment failed, there is already a duplicated transaction fee applied to the cart summary (on the very right in the screen capture) --> this is still the case and needs fixing
The transaction fee is still in the cart (see screen capture) --> this needs fixing

I would not say these need fixing, let's perhaps discuss prioritization if we choose to open an issue?

I'd say there are minor issues (s4?) and only occurring for failed Paypal payment attempts, specifically with a bad signature. Does this occur with a Stripe failed payments as well? Or a Paypal insufficient funds? Seems out of scope for this PR, which effectively  prevents the piling up of transaction fees after the user attempts to start a new order, by emptying the cart. What do you think?

Merging! :tada:

**Comment by drummer83** [comment]
Thanks for checking @filipefurtad0!

Agree, that these would be minor and separate issues. I will definitely create one for the order summary, because my understanding of math can't ignore this:
grafik

Were you able to test this with failing Stripe payment as well? I couldn't even reproduce the original issue (https://github.com/openfoodfoundation/openfoodnetwork/issues/9112#issuecomment-1106439837).

**Comment by filipefurtad0** [comment]
> definitely create one for the order summary
Thanks. Yes, this is surely confusing to users, but to my understanding, an edge case.

> test this with failing Stripe payment as well
Nope, I could not reproduce it - neither before or after the PR. We need to keep an eye for multiple transaction fees.

**Pull request merged**